File: api/internal/utils/utils_test.go
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package utils

import (
	"errors"
	"os"
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestGetFlakeUid(t *testing.T) {
	id := GetFlakeUid()
	assert.NotEqual(t, 0, id)
}

func TestGetFlakeUidStr(t *testing.T) {
	id := GetFlakeUidStr()
	assert.NotEqual(t, "", id)
	assert.Equal(t, 18, len(id))
}

func TestGetLocalIPs(t *testing.T) {
	_, err := getLocalIPs()
	assert.Equal(t, nil, err)
}

func TestSumIPs_with_nil(t *testing.T) {
	total := sumIPs(nil)
	assert.Equal(t, uint16(0), total)
}

func TestObjectClone(t *testing.T) {
	type test struct {
		Str string
		Num int
	}

	origin := &test{Str: "a", Num: 1}
	copy := &test{}
	err := ObjectClone(origin, copy)
	assert.Nil(t, err)
	assert.Equal(t, origin, copy)

	// change value of the copy, should not change value of origin
	copy.Num = 2
	assert.NotEqual(t, copy.Num, origin.Num)
	assert.Equal(t, 1, origin.Num)
}

func TestGenLabelMap(t *testing.T) {
	expectedErr := errors.New("malformed label")
	mp, err := GenLabelMap("l1")
	assert.Nil(t, err)
	assert.Equal(t, mp["l1"], struct{}{})

	mp, err = GenLabelMap("l1,l2:v2")
	assert.Nil(t, err)
	assert.Equal(t, mp["l1"], struct{}{})
	assert.Equal(t, mp["l2:v2"], struct{}{})

	mp, err = GenLabelMap("l1:v1,l1:v2")
	assert.Nil(t, err)
	assert.Equal(t, mp["l1:v1"], struct{}{})
	assert.Equal(t, mp["l1:v2"], struct{}{})

	mp, err = GenLabelMap(",")
	assert.Equal(t, expectedErr, err)
	assert.Nil(t, mp)

	mp, err = GenLabelMap(",l2:,")
	assert.Equal(t, expectedErr, err)
	assert.Nil(t, mp)
}

func TestLabelContains(t *testing.T) {
	reqMap, _ := GenLabelMap("l1,l2:v2")
	mp := map[string]string{
		"l1": "v1",
	}
	assert.True(t, LabelContains(mp, reqMap))

	mp = map[string]string{
		"l1": "v1",
		"l2": "v3",
	}
	assert.True(t, LabelContains(mp, reqMap))

	mp = map[string]string{
		"l2": "v3",
	}
	assert.False(t, LabelContains(mp, reqMap))

	reqMap, _ = GenLabelMap("l1:v1,l1:v2")
	mp = map[string]string{
		"l1": "v1",
	}
	assert.True(t, LabelContains(mp, reqMap))

	reqMap, _ = GenLabelMap("l1:v1,l1:v2")
	mp = map[string]string{
		"l1": "v2",
	}
	assert.True(t, LabelContains(mp, reqMap))
}

func TestValidateLuaCode(t *testing.T) {
	validLuaCode := "local _M = {} \n function _M.access(api_ctx) \n ngx.log(ngx.WARN,\"hit access phase\") \n end \nreturn _M"
	err := ValidateLuaCode(validLuaCode)
	assert.Nil(t, err)

	invalidLuaCode := "local _M = {} \n function _M.access(api_ctx) \n ngx.log(ngx.WARN,\"hit access phase\")"
	err = ValidateLuaCode(invalidLuaCode)
	assert.NotNil(t, err)
	assert.Equal(t, "<string> at EOF:   syntax error\n", err.Error())
}

func TestPathExist(t *testing.T) {
	workDir, _ := os.Getwd()
	exist, err := PathExist(workDir)
	assert.Nil(t, err)
	assert.True(t, exist)
}

File: src/components/schema-form/__tests__/utils.test.ts
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { describe, expect, it } from 'vitest';

import type { JSONSchema } from '../types';
import {
  evaluateIfCondition,
  fieldPath,
  getFieldLabel,
  getOneOfLabel,
  getOneOfMatchIndex,
  getRequiredSet,
  getSchemaType,
  hasConditionals,
  hasDependencies,
  isEncryptField,
  isEnumField,
  isRequired,
  mergeSchemas,
  resolveDefaults,
} from '../utils';

describe('getSchemaType', () => {
  it('returns the type string directly', () => {
    expect(getSchemaType({ type: 'string' })).toBe('string');
    expect(getSchemaType({ type: 'integer' })).toBe('integer');
    expect(getSchemaType({ type: 'boolean' })).toBe('boolean');
  });

  it('returns first type from array', () => {
    expect(getSchemaType({ type: ['string', 'null'] })).toBe('string');
  });

  it('infers type from enum', () => {
    expect(getSchemaType({ enum: ['a', 'b'] })).toBe('string');
  });

  it('infers object from properties', () => {
    expect(getSchemaType({ properties: { foo: { type: 'string' } } })).toBe('object');
  });

  it('infers array from items', () => {
    expect(getSchemaType({ items: { type: 'string' } })).toBe('array');
  });

  it('infers oneOf', () => {
    expect(getSchemaType({ oneOf: [{ type: 'string' }] })).toBe('oneOf');
  });

  it('returns undefined for empty schema', () => {
    expect(getSchemaType({})).toBeUndefined();
  });
});

describe('isEnumField', () => {
  it('returns true for schema with enum', () => {
    expect(isEnumField({ enum: ['a', 'b', 'c'] })).toBe(true);
  });

  it('returns false for schema without enum', () => {
    expect(isEnumField({ type: 'string' })).toBe(false);
  });

  it('returns false for empty enum', () => {
    expect(isEnumField({ enum: [] })).toBe(false);
  });
});

describe('isEncryptField', () => {
  it('returns true when field name is in encrypt list', () => {
    expect(isEncryptField('client_secret', ['client_secret', 'client_key'])).toBe(true);
  });

  it('handles dotted paths by checking last segment', () => {
    expect(isEncryptField('auth.client_secret', ['client_secret'])).toBe(true);
  });

  it('returns false when not in list', () => {
    expect(isEncryptField('client_id', ['client_secret'])).toBe(false);
  });

  it('returns false for undefined list', () => {
    expect(isEncryptField('secret', undefined)).toBe(false);
  });
});

describe('isRequired', () => {
  it('returns true when field is in required array', () => {
    const schema: JSONSchema = { type: 'object', required: ['name', 'key'] };
    expect(isRequired('name', schema)).toBe(true);
  });

  it('returns false when field is not in required array', () => {
    const schema: JSONSchema = { type: 'object', required: ['name'] };
    expect(isRequired('key', schema)).toBe(false);
  });

  it('returns false when no required array', () => {
    expect(isRequired('name', { type: 'object' })).toBe(false);
  });
});

describe('getRequiredSet', () => {
  it('returns a Set of required fields', () => {
    const schema: JSONSchema = { required: ['a', 'b'] };
    const set = getRequiredSet(schema);
    expect(set.has('a')).toBe(true);
    expect(set.has('b')).toBe(true);
    expect(set.has('c')).toBe(false);
  });

  it('returns empty Set for no required', () => {
    expect(getRequiredSet({}).size).toBe(0);
  });
});

describe('resolveDefaults', () => {
  it('extracts defaults from properties', () => {
    const schema: JSONSchema = {
      type: 'object',
      properties: {
        policy: { type: 'string', default: 'local' },
        timeout: { type: 'integer', default: 3 },
        name: { type: 'string' },
      },
    };
    const defaults = resolveDefaults(schema);
    expect(defaults).toEqual({ policy: 'local', timeout: 3 });
  });

  it('handles nested objects', () => {
    const schema: JSONSchema = {
      type: 'object',
      properties: {
        session: {
          type: 'object',
          properties: {
            storage: { type: 'string', default: 'cookie' },
          },
        },
      },
    };
    const defaults = resolveDefaults(schema);
    expect(defaults).toEqual({ session: { storage: 'cookie' } });
  });

  it('returns empty for non-object schemas', () => {
    expect(resolveDefaults({ type: 'string' })).toEqual({});
  });
});

describe('fieldPath', () => {
  it('combines root and field name', () => {
    expect(fieldPath('config', 'timeout')).toBe('config.timeout');
  });

  it('returns field name when root is empty', () => {
    expect(fieldPath('', 'timeout')).toBe('timeout');
  });
});

describe('getFieldLabel', () => {
  it('uses schema title when available', () => {
    expect(getFieldLabel('foo', { title: 'My Field' })).toBe('My Field');
  });

  it('generates label from field name', () => {
    expect(getFieldLabel('redis_host', {})).toBe('Redis host');
  });

  it('handles camelCase', () => {
    expect(getFieldLabel('clientSecret', {})).toBe('Client Secret');
  });

  it('handles dotted paths', () => {
    expect(getFieldLabel('config.redis_host', {})).toBe('Redis host');
  });
});

describe('hasConditionals', () => {
  it('returns true for if/then', () => {
    expect(hasConditionals({ if: { properties: {} }, then: {} })).toBe(true);
  });

  it('returns false for no conditionals', () => {
    expect(hasConditionals({ type: 'object' })).toBe(false);
  });
});

describe('hasDependencies', () => {
  it('returns true when dependencies exist', () => {
    expect(hasDependencies({ dependencies: { a: ['b'] } })).toBe(true);
  });

  it('returns false for empty', () => {
    expect(hasDependencies({})).toBe(false);
  });
});

describe('evaluateIfCondition', () => {
  it('returns then when enum matches', () => {
    const schema: JSONSchema = {
      if: { properties: { policy: { enum: ['redis'] } } },
      then: { required: ['redis_host'] },
      else: {},
    };
    expect(evaluateIfCondition(schema, { policy: 'redis' })).toBe('then');
  });

  it('returns else when enum does not match', () => {
    const schema: JSONSchema = {
      if: { properties: { policy: { enum: ['redis'] } } },
      then: { required: ['redis_host'] },
      else: {},
    };
    expect(evaluateIfCondition(schema, { policy: 'local' })).toBe('else');
  });

  it('returns null when no if condition', () => {
    expect(evaluateIfCondition({}, {})).toBeNull();
  });
});

describe('mergeSchemas', () => {
  it('merges properties from two schemas', () => {
    const base: JSONSchema = {
      type: 'object',
      properties: { a: { type: 'string' } },
    };
    const ext: JSONSchema = {
      properties: { b: { type: 'integer' } },
      required: ['b'],
    };
    const merged = mergeSchemas(base, ext);
    expect(merged.properties?.a).toBeDefined();
    expect(merged.properties?.b).toBeDefined();
    expect(merged.required).toContain('b');
  });
});

describe('getOneOfMatchIndex', () => {
  it('matches by required fields', () => {
    const schemas: JSONSchema[] = [
      { required: ['conn', 'burst'] },
      { required: ['rules'] },
    ];
    expect(getOneOfMatchIndex(schemas, { conn: 10, burst: 5 })).toBe(0);
    expect(getOneOfMatchIndex(schemas, { rules: [{}] })).toBe(1);
  });

  it('defaults to 0 when no match', () => {
    const schemas: JSONSchema[] = [
      { required: ['a'] },
      { required: ['b'] },
    ];
    expect(getOneOfMatchIndex(schemas, {})).toBe(0);
  });
});

describe('getOneOfLabel', () => {
  it('uses title when available', () => {
    expect(getOneOfLabel({ title: 'Redis Mode' }, 0)).toBe('Redis Mode');
  });

  it('uses required fields as label', () => {
    expect(getOneOfLabel({ required: ['conn', 'burst'] }, 0)).toBe('conn + burst');
  });

  it('falls back to Option N', () => {
    expect(getOneOfLabel({}, 2)).toBe('Option 3');
  });
});
